File: Coin Change 2/SpaceOptimization.py
class Solution:

    def change(self, amount: int, coins: List[int]) -> int:

        size = len(coins)
        prev =  [0 for _ in range(amount+1)] 
        cur =  [0 for _ in range(amount+1)] 

        for index in range(size):
            for target in range(amount+1):
                if target == 0: cur[0] = 1
                else:
                    take = 0
                    if coins[index] <= target: take = cur[target-coins[index]]
                    nonTake = prev[target]
                    cur[target] =  take + nonTake

            prev = cur

        return prev[amount]

File: Coin Change 2/SpaceOptimization.java
class SpaceOptimization {


    public int change(int amount, int[] coins) {
        int size = coins.length;
        int[] prev = new int[amount+1];
        int[] cur = new int[amount+1];

        for(int index = 0 ; index < size; index++) {
            for(int target = 0 ;target <= amount; target ++) {
                if(target == 0) cur[target] = 1;
                else {
                    int take = 0;
                    if(coins[index] <=  target) 
                        take = cur[target-coins[index]];
                    
                    int nonTake = 0;
                    if(index >= 1) nonTake =  prev[target];
                    cur[target] = nonTake + take;
                }
            }

            prev = cur;
        }

        return prev[amount];

    }
}